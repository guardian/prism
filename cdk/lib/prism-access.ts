import { GuArnParameter, GuStack } from '@guardian/cdk/lib/constructs/core';
import { AppIdentity } from '@guardian/cdk/lib/constructs/core/identity';
import { GuRole } from '@guardian/cdk/lib/constructs/iam';
import { CfnOutput } from 'aws-cdk-lib';
import type { App } from 'aws-cdk-lib';
import {
	ArnPrincipal,
	Effect,
	Policy,
	PolicyStatement,
} from 'aws-cdk-lib/aws-iam';

export class PrismAccess extends GuStack {
	private static app: AppIdentity = {
		app: 'prism',
	};

	constructor(scope: App, id: string) {
		super(scope, id, {
			description: 'CloudFormation template to create the prism role.',
			stack: 'deploy',
			stage: 'INFRA', // singleton stack,
			env: {
				region: 'eu-west-1',
			},
		});

		/*
    Looks like some @guardian/cdk constructs are not applying the App tag.
    I suspect since https://github.com/guardian/cdk/pull/326.
    Until that is fixed, we can safely, manually apply it to all constructs in tree from `this` as it's a single app stack.
    TODO: remove this once @guardian/cdk has been fixed.
     */
		AppIdentity.taggedConstruct(PrismAccess.app, this);

		const parameters = {
			PrismAccount: new GuArnParameter(this, 'PrismAccount', {
				description:
					'The ARN of the account in which Prism is running - looks like arn:aws:iam::<account-number>:root',
			}),
		};

		/*
		 * This is the external prism role in each account which is used by prism to crawl data from that account.
		 */
		const prismRole = new GuRole(this, 'PrismRole', {
			description: 'Role Prism uses to crawl resources in this account',
			assumedBy: new ArnPrincipal(parameters.PrismAccount.valueAsString),
		});

		this.overrideLogicalId(prismRole, {
			logicalId: 'PrismRole',
			reason:
				'We override this to ensure that we do not replace the existing resource',
		});

		new Policy(this, 'PrismPolicy', {
			policyName: 'PrismCollection',
			roles: [prismRole],
			statements: [
				new PolicyStatement({
					effect: Effect.ALLOW,
					resources: ['*'],
					actions: [
						'ec2:Describe*',
						'iam:Get*',
						'iam:List*',
						'autoscaling:Describe*',
						's3:ListAllMyBuckets',
						's3:GetBucketLocation',
						'route53:List*',
						'route53:Get*',
						'elasticloadbalancing:Describe*',
						'lambda:ListFunctions',
						'lambda:ListTags',
						'cloudformation:Describe*',
						'cloudformation:Get*',
					],
				}),
			],
		});

		new CfnOutput(this, 'Role', {
			value: prismRole.roleArn,
			description: 'Prism Role',
		});
	}
}
