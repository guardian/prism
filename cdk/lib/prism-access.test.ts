import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { PrismAccess } from './prism-access';

describe('The PrismAccess stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new PrismAccess(app, 'prism-access');

		expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
	});
});
