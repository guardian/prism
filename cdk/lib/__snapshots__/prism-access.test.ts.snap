// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The PrismAccess stack matches the snapshot 1`] = `
Object {
  "Description": "CloudFormation template to create the prism role.",
  "Outputs": Object {
    "Role": Object {
      "Description": "Prism Role",
      "Value": Object {
        "Fn::GetAtt": Array [
          "PrismRole",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "PrismAccount": Object {
      "AllowedPattern": "arn:aws:[a-z0-9]*:[a-z0-9\\\\-]*:[0-9]{12}:.*",
      "ConstraintDescription": "Must be a valid ARN, eg: arn:partition:service:region:account-id:resource-id",
      "Description": "The ARN of the account in which Prism is running - looks like arn:aws:iam::<account-number>:root",
      "Type": "String",
    },
  },
  "Resources": Object {
    "PrismPolicyF951F5B3": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2:Describe*",
                "iam:Get*",
                "iam:List*",
                "autoscaling:Describe*",
                "s3:ListAllMyBuckets",
                "s3:GetBucketLocation",
                "acm:ListCertificates",
                "acm:DescribeCertificate",
                "route53:List*",
                "route53:Get*",
                "elasticloadbalancing:Describe*",
                "lambda:ListFunctions",
                "lambda:ListTags",
                "rds:Describe*",
                "cloudformation:Describe*",
                "cloudformation:Get*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PrismCollection",
        "Roles": Array [
          Object {
            "Ref": "PrismRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PrismRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Ref": "PrismAccount",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role Prism uses to crawl resources in this account",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "prism",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/prism",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "INFRA",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
